pylint/checkers/len.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (c) 2016 Alexander Todorov <atodorov@MrSenko.com>
+
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/master/COPYING
+
+"""Check for incorrect usage of len() in conditions."""
+import itertools
+import astroid
+
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import check_messages
+
+MSGS = {
+    'W1801': ('Do not use `len(SEQUENCE)` as condition value',
+              'len-as-condition',
+              'Used when Pylint detects incorrect use of len(sequence) inside '
+              'conditions.'),
+    }
+
+
+def _is_len_call(node):
+    """Checks if @node is len(SOMETHING)."""
+    return (isinstance(node, astroid.Call) and isinstance(node.func, astroid.Name) and
+            node.func.name == 'len')
+
+def _is_constant_zero(node):
+    return isinstance(node, astroid.Const) and (node.value == 0)
+
+
+class LenChecker(BaseChecker):
+    """Checks for incorrect usage of len() inside conditions. pep8 states:
+    For sequences, (strings, lists, tuples), use the fact that empty sequences are false.
+
+        Yes: if not seq:
+             if seq:
+
+        No: if len(seq):
+            if not len(seq):
+
+    Problems detected:
+    * if len(sequence):
+    * if not len(sequence):
+    * if len(sequence) == 0:
+    * if len(sequence) != 0:
+    * if len(sequence) > 0:
+    """
+
+    __implements__ = (IAstroidChecker,)
+
+    # configuration section name
+    name = 'len'
+    # messages
+    msgs = MSGS
+    priority = -2
+    # configuration options
+    options = ()
+
+    @check_messages('len-as-condition')
+    def visit_call(self, node):
+        # a len(S) call is used inside a test condition
+        # could be if, while, assert or if expression statement
+        # e.g. `if len(S):`
+        if _is_len_call(node):
+            # the len() call could also be nested together with other
+            # boolean operations, e.g. `if z or len(x):`
+            parent = node.parent
+            while isinstance(parent, astroid.BoolOp):
+                parent = parent.parent
+
+            # we're finally out of any nested boolean operations so check if
+            # the parent is indeed an if, while, assert of if expression statement
+            if hasattr(parent, 'test'):
```

Comment:
I prefer an explicit instance check, something as in ``isinstance(parent, (astroid.If, ...))```